### ISSUE TITLE
Alternate props syntax

### ISSUE THREAD

**Issue by za-creature** [yyx990803/vue#978]
Currently, properties are defined as an array of names:

```javascript
Vue.component("test", {
    "props": ["hello", "world"]
})
```

Or as an array of objects:

```javascript
Vue.component("test", {
    "props": [{
        "name": "hello",
        "type": String
    }, {
        "name": "world",
        "type": Boolean
    }]
})
```

While the first syntax is okay for enumerating properties, the second one is a bit awkward and inconsistent with the way `data` entries are specified.

I propose the following syntax extensions inspired from [mongoose](http://mongoosejs.com/):

```javascript
Vue.component("test", {
    "props": {
        "hello": String,
        "world": {
            "type": Boolean,
            "required": false
        }
})
```

I believe this could be implemented without breaking compatibility.

**Comment by okoala** [comment]
:+1:

**Comment by butsalt** [comment]
Hmm... I'm better like current props syntax.
First, an array of Object and an array of String are more identical in format, because both of them are Array.
Second, I don't know why the value means type configuration to target property rather than any other configuration like validate. I doubt that will be a bit confusing...

**Comment by okoala** [comment]
key value的形式可读性比较高，看起来比较简洁~~

**Comment by za-creature** [comment]
@butsalt I was talking about consistency with the `data` attribute, which is a key-value map. Arrays of String work well and have excellent readability when you just want to enumerate a list of properties, but if you also want to add options to them, reading the list becomes difficult imo.

As for the `"hello": String` thing, it's just a shortcut designed to condense stuff like:

```
"props": {
    "firstName": {
        "type": String
    },
    "lastName": {
        "type": String
    },
    "gender": {
        "type": String
    }
    "email": {
        "type": String
    },
    "age": {
        "type": Number
    }
}
```

into a more easy to read:

```
"props": {
    "firstName": String,
    "lastName": String,
    "gender": String,
    "email": String,
    "age": Number
}
```

**Comment by yyx990803** [comment]
Yes, the key-value map style is more consistent with the other options. I used the Array<Object> syntax mostly because it was extended from the original Array<String> processing logic.

It's landed in 81911b4491b9e328d0a04e7f5589fcac985af9c0 without breaking back-compat.
